src/Compiler/Driver/OptimizeInputs.fs

```diff
@@ -48,6 +53,241 @@ let GetInitialOptimizationEnv (tcImports: TcImports, tcGlobals: TcGlobals) =
     let optEnv = List.fold (AddExternalCcuToOptimizationEnv tcGlobals) optEnv ccuinfos
     optEnv
 
+type private OptimizeDuringCodeGen = bool -> Expr -> Expr
+
+type PhaseContext =
+    {
+        OptEnvFirstLoop: Optimizer.IncrementalOptimizationEnv
+        OptInfo: Optimizer.ImplFileOptimizationInfo
+        OptEnvExtraLoop: Optimizer.IncrementalOptimizationEnv
+        OptEnvFinalSimplify: Optimizer.IncrementalOptimizationEnv
+        HidingInfo: SignatureHidingInfo
+        OptDuringCodeGen: OptimizeDuringCodeGen
+    }
+
+type PhaseRes = CheckedImplFile * PhaseContext
+
+type PhaseIdx = int
+type F<'In, 'Out> = 'In -> 'Out
+
+type PhaseInputs =
+    {
+        File: CheckedImplFile
+        FileIdx: int
+        // State returned by processing the previous phase for the current file.
+        PrevPhase: PhaseContext
+        // State returned by processing the current phase for the previous file.
+        PrevFile: PhaseContext
+    }
+
+type PhaseFunc = PhaseInputs -> CheckedImplFile * PhaseContext
+
+/// <summary>
+/// Each file's optimization can be split into three different phases, executed one after another.
+/// Each phase calls 'Optimizer.OptimizeImplFile' and performs some other tasks.
+/// Each phase uses outputs of the previous phase and outputs of previous file's optimization for the same phase.
+/// </summary>
+type Phase =
+    {
+        Idx: PhaseIdx
+        Name: string
+    }
+
+    override this.ToString() = $"{this.Idx}-{this.Name}"
+
+type PhaseInfo = { Phase: Phase; Func: PhaseFunc }
+
+type PhaseInfos = PhaseInfo[]
+
+[<RequireQualifiedAccess>]
+module private ParallelOptimization =
+    open Optimizer
+
+    /// Identifies a work item scheduled independently of others - consists of a (file) index and OptimizationPhase.
+    /// There are (NumberOfFiles * NumberOfPhases) nodes in the whole optimization process.
+    type private Node =
+        {
+            FileIdx: int
+            Phase: PhaseIdx
+        }
+
+        override this.ToString() = $"[{this.FileIdx}-{this.Phase}]"
+
+    /// Final processing of file results to produce output needed for further compilation steps.
+    let private collectFinalResults
+        (fileResults: PhaseRes[])
+        : (CheckedImplFileAfterOptimization * ImplFileOptimizationInfo)[] * IncrementalOptimizationEnv =
+        let finalFileResults =
+            fileResults
+            |> Array.map (fun (file, res) ->
+                let implFile =
+                    {
+                        ImplFile = file
+                        OptimizeDuringCodeGen = res.OptDuringCodeGen
+                    }
+
+                implFile, res.OptInfo)
+
+        let lastFileFirstLoopEnv =
+            fileResults |> Array.last |> (fun (_file, res) -> res.OptEnvFirstLoop)
+
+        finalFileResults, lastFileFirstLoopEnv
+
+    let private raiseNoResultsExn (node: Node) =
```

Comment:
Is this getting called from somewhere?
Maybe better as a debug assert() ?

Reply:
Not anymore, after changing to a task-based approach.